modules/ssh/ssh.go

```diff
@@ -10,178 +9,179 @@ import (
-					log.Trace("SSH: Arguments: %v", args)
-					cmd := exec.Command(setting.AppPath, args...)
-					cmd.Env = append(
-						os.Environ(),
-						"SSH_ORIGINAL_COMMAND="+cmdName,
-						"SKIP_MINWINSVC=1",
-					)
-
-					stdout, err := cmd.StdoutPipe()
-					if err != nil {
-						log.Error("SSH: StdoutPipe: %v", err)
-						return
-					}
-					stderr, err := cmd.StderrPipe()
-					if err != nil {
-						log.Error("SSH: StderrPipe: %v", err)
-						return
-					}
-					input, err := cmd.StdinPipe()
-					if err != nil {
-						log.Error("SSH: StdinPipe: %v", err)
-						return
-					}
-
-					// FIXME: check timeout
-					if err = cmd.Start(); err != nil {
-						log.Error("SSH: Start: %v", err)
-						return
-					}
-
-					err = req.Reply(true, nil)
-					if err != nil {
-						log.Error("SSH: Reply: %v", err)
-					}
-					go func() {
-						_, err = io.Copy(input, ch)
-						if err != nil {
-							log.Error("SSH: Copy: %v", err)
-						}
-					}()
-					_, err = io.Copy(ch, stdout)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-					_, err = io.Copy(ch.Stderr(), stderr)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-
-					if err = cmd.Wait(); err != nil {
-						log.Error("SSH: Wait: %v", err)
-						return
-					}
-
-					_, err = ch.SendRequest("exit-status", false, []byte{0, 0, 0, 0})
-					if err != nil {
-						log.Error("SSH: SendRequest: %v", err)
-					}
-					return
-				default:
-				}
-			}
-		}(reqs)
+		return 1
 	}
+
+	return waitStatus.ExitStatus()
 }
 
-func listen(config *ssh.ServerConfig, host string, port int) {
-	listener, err := net.Listen("tcp", host+":"+com.ToStr(port))
+func shellEscape(args []string) string {
+	var data = make([]string, len(args))
+	for k, v := range args {
+		data[k] = v
+
+		// This is a borderline dirty hack. It's designed to only escape
+		// strings which *might* need to be escaped. It uses a very
+		// limited character set so everything that needs to be quoted
+		// will be but we can ignore some simple cases to make it easier
+		// to parse in the ssh hook. We allow alpha-numeric characters
+		// along with ., _, and -
+		idx := strings.IndexFunc(v, func(r rune) bool {
+			return !unicode.IsLetter(r) && !unicode.IsDigit(r) && !strings.ContainsRune(".-_", r)
+		})
+		if idx > -1 || v == "" {
+			data[k] = "'" + strings.Replace(data[k], "'", "'\"'\"'", -1) + "'"
+		}
+	}
+	return strings.Join(data, " ")
+}
+
+func sessionHandler(session ssh.Session) {
+	keyID := session.Context().Value(giteaKeyID).(int64)
+
+	command := shellEscape(session.Command())
+
+	log.Trace("SSH: Payload: %v", command)
+
+	args := []string{"serv", "key-" + com.ToStr(keyID), "--config=" + setting.CustomConf}
+	log.Trace("SSH: Arguments: %v", args)
+	cmd := exec.Command(setting.AppPath, args...)
+	cmd.Env = append(
+		os.Environ(),
+		"SSH_ORIGINAL_COMMAND="+command,
+		"SKIP_MINWINSVC=1",
+	)
+
+	stdout, err := cmd.StdoutPipe()
```

Comment:
I haven't checked this completely but can't we just put the session directly in to cmd as Stdout, stdin and stderr?

Reply:
Unfortunately, no. Switching to using them directly would introduce a race condition. We have to wait for stdout and stderr to finish sending.

Reply:
Fair enough. It just seems wasteful to create 6 extra goroutines with three extra buffers.

I wonder if we could get away with wrapping the writeClosers with something that `defer wg.Done()` on `Close()`?

Reply:
That's an interesting idea - I went with this route because it's still relatively easy to understand, but I think your suggestion would also work (as long as we did something similar with stdin causing it to actually close when done copying).

In short, running a command needs these (in order):

- Run the command
- Wait for command to exit (note that some git commands will not exit until stdin is manually closed, as we do here when the io.Copy is done)
- Wait for write buffers to be written
- Return exit status

Reply:
I guess the question is do we always receive a close from cmd no matter what error is received in write and are closes always deferred.

Reply:
Maybe I'm misunderstanding, but how does this make 6 goroutines?

Reply:
I would strongly recommend waiting for a command to finish, but it depends on your needs. Someone could abuse the server by starting a bunch of commands and disconnecting. However, we don't want to just kill git processes (this way we don't kill it mid-way through a repack) when the user disconnects. In theory the closing of stdin when it's closed remotely should take care of the worst of this.

Reply:
It's no problem. I just wondered if we could avoid it. (io.pipe has a goroutine internally)

Reply:
We probably could - it would be possible to improve this, but it's still an improvement over the current code. This will also still be lighter than openssh, as by default it forks for every connection (so if a connection dies, it doesn't take out the whole server)